src/systems/environmental_sensor_system/EnvironmentalSensorSystem.cc

```diff
@@ -0,0 +1,349 @@
+#include <gz/sim/components/ParentEntity.hh>
+#include <gz/sim/components/Name.hh>
+#include <gz/sim/components/Environment.hh>
+#include <gz/sim/Util.hh>
+
+#include <gz/sensors/SensorFactory.hh>
+#include <gz/sensors/SensorTypes.hh>
+#include <gz/sensors/Util.hh>
+
+#include <gz/math/Pose3.hh>
+
+#include <chrono>
+#include <unordered_map>
+#include <utility>
+#include <unordered_set>
+#include <memory>
+#include <string>
+
+using namespace gz;
+using namespace gz::sim;
+
+/// Sensor prefix to be used. All envionment_sensors are to be prefixed by
+/// "environment_sensor/" in their gz:type field.
+const char* SENSOR_TYPE_PREFIX =  "environmental_sensor/";
+
+////////////////////////////////////////////////////////////////
+/// \brief Envirtonment Sensor used for looking up environment values in our
+/// CSV file.
+class EnvironmentalSensor : public gz::sensors::Sensor
+{
+  /// \brief Node for communication
+  private: gz::transport::Node node;
+
+  /// \brief Publishes sensor data
+  private: gz::transport::Node::Publisher pub;
+
+  ////////////////////////////////////////////////////////////////
+  /// Documentation inherited
+  public: virtual bool Load(const sdf::Sensor &_sdf) override
+  {
+    auto type = gz::sensors::customType(_sdf);
+    if (type.substr(0, strlen(SENSOR_TYPE_PREFIX)) != SENSOR_TYPE_PREFIX ||
+      type.size() <= strlen(SENSOR_TYPE_PREFIX))
+    {
+      gzerr << "Trying to load [" << SENSOR_TYPE_PREFIX
+        << "] sensor, but got type ["
+        << type << "] instead." <<  std::endl;
+      return false;
+    }
+
+    // Load common sensor params
+    gz::sensors::Sensor::Load(_sdf);
+
+    this->pub = this->node.Advertise<gz::msgs::Double>(this->Topic());
+
+    this->field = type.substr(strlen(SENSOR_TYPE_PREFIX));
+
+    gzdbg << "Loaded environmental sensor for " << this->field
+      << " publishing on " << this->Topic() << std::endl;
+
+    return true;
+  }
+
+  ////////////////////////////////////////////////////////////////
+  /// \brief Update the sensor and generate data
+  /// \param[in] _now The current time
+  /// \return True if the update was successful
+  public: virtual bool Update(
+    const std::chrono::steady_clock::duration &_now) override
+  {
+    if (!this->ready) return false;
+
+    if (!this->session.has_value()) return false;
+
+    this->session = this->gridField->frame[this->field].StepTo(
+      this->session.value(), std::chrono::duration<double>(_now).count());
+
+    if (!this->session.has_value()) return false;
+
+    gz::msgs::Double msg;
+    *msg.mutable_header()->mutable_stamp() = gz::msgs::Convert(_now);
+    auto frame = msg.mutable_header()->add_data();
+    frame->set_key("frame_id");
+    frame->add_value(this->Name());
+    auto data = this->gridField->frame[this->field].LookUp(
+      this->session.value(), this->position);
+    if (!data.has_value())
+      return false;
+    msg.set_data(data.value());
+    // TODO(anyone) Add sensor noise.
```

Comment:
@arjo129 fyi: should be super straightforward using [`gz::sensors::NoiseFactory`](https://github.com/gazebosim/gz-sensors/blob/bbdb620d0ab9c3cc4331e8006f1ce1fa79f9bb2d/include/ignition/sensors/Noise.hh#L42).